### ISSUE TITLE
How can I do 3-dimensional regression?

### ISSUE THREAD

**Issue by ByungtaeAhn** [vlfeat/matconvnet#102]
Hi,

I'm trying to reproduce the method described in the paper:
"Real-time Head Orientation from a Monocular Camera using Deep Neural Network, ACCV2014"

The method estimates head orientation in a 3-dimensional regression manner. How can I do the regression with matconvnet? I think I should make L2 norm loss function first, and modify the source code for 3-dimensional backpropagation.

Here is my net structure.

=========================================================
net.layers = {} ;
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', f*randn(5,5,1,16, 'single'), ...
                           'biases', zeros(1, 16, 'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;
net.layers{end+1} = struct('type', 'pool', ...
                           'method', 'max', ...
                           'pool', [2 2], ...
                           'stride', 2, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', f*randn(5,5,16,20, 'single'),...
                           'biases', zeros(1,20,'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;
net.layers{end+1} = struct('type', 'pool', ...
                           'method', 'max', ...
                           'pool', [2 2], ...
                           'stride', 2, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', f*randn(3,3,20,120, 'single'),...
                           'biases', zeros(1,120,'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', f*randn(3,3,120,84, 'single'),...
                           'biases', zeros(1,84,'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', f*randn(1,1,84,3, 'single'),...
                           'biases', zeros(1,3,'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;

net.layers{end+1} = struct('type', 'l2normloss') ;

=========================================================
function Y = vl_l2normloss(X,c,dzdy)

if nargin <= 2 
  Y = 0.5*sum((squeeze(X)'-c').^2);
else
  Y = +((squeeze(X)'-c'))*dzdy;
  Y = reshape(Y,size(X));
end

=========================================================

The last layer has 3 neurons, and the size of X,c is 3xN.
However, when the net starts training, it outputs the error that the dimensions are not matched each other at cnn_train->updateError (line 273)

How can I implement for the 3-dimensional backpropagation?
@lenck @vedaldi  Any insight on what might seems to be the problem?

Thanks!
Byungtae

**Comment by lenck** [comment]
Hi, sorry for late response...
The l2norm_loss seems fine, only thing is that you don't normalise it by the number of data points which may lead to different effective learning rate when you would change the batch size...
e.g. I wrote the squared loss this way:
```
assert(numel(X) == numel(C));

if nargin <= 2
  Y = sum((X(:) - C(:)).^2) ./ (2*numel(X));
else
  assert(numel(dzdy) == 1);
  Y = reshape((dzdy / numel(X)) * (X(:) - C(:)), size(X));
end
```
And it seems to work for me :) But yours is basically identical...
And with the updateError function - just simply ignore it and for example write your own error type... The update error in matconvnet is written for softmax only right now... :P I've just added there:
```
  case 'squaredist'
    n = prod(sz(1:2)) ;
    error = sum((predictions(:) - labels(:)).^2) ;
    info.error(end) = info.error(end) + error/n ;
```
When I needed the L2 regression :) Though I'm not 100% sure if it is correct..

**Comment by AnanS** [comment]
@lenck 

Doesn't cnn_train already divide the gradient by the batch size. So, if you use the above loss function, you would end up dividing by the batch size twice (I was dividing the first Y by numel(X) as above but the second Y by size(X, 1) * size(X, 2) * size(X, 3) only rather than numel(X)). Or am I missing something?

And I guess '2 *' should be '/ 2'.

Thanks!

**Comment by lenck** [comment]
Yes, you are right, there was a bug :) I was wondering why the training depends so much on the batch size ;) Thanks for pointing that out!
So I think it should be like this (as cnn_train does not divide by the number of output cells):
```
assert(numel(X) == numel(C));
n = size(X,1) * size(X,2);
if nargin <= 2
  Y = sum((X(:) - C(:)).^2) ./ (2*n);
else
  assert(numel(dzdy) == 1);
  Y = reshape((dzdy / n) * (X(:) - C(:)), size(X));
end
```

**Comment by anurikadisha** [comment]
Hi @AnanS 
I also want to change the softmax loss function. In this file they do not divide by the batch size. Can you please let me know where is the loss divided by the batch size in cnn_train?

The following is the code for softmax loss:
sz = [size(X,1) size(X,2) size(X,3) size(X,4)] ;

% index from 0
c = c - 1 ;

if numel(c) == sz(4)
  % one label per image
  c = reshape(c, [1 1 1 sz(4)]) ;
  c = repmat(c, [sz(1) sz(2)]) ;
else
  % one label per spatial location
  sz_ = [size(c,1) size(c,2) size(c,3) size(c,4)] ;
  assert(isequal(sz_, [sz(1) sz(2) 1 sz(4)])) ;
end

% convert to indeces
c_ = 0:numel(c)-1 ;
c_ = 1 + ...
  mod(c_, sz(1)*sz(2)) + ...
  (sz(1)*sz(2)) * c(:)' + ...
  (sz(1)*sz(2)*sz(3)) * floor(c_/(sz(1)*sz(2))) ;

% compute softmaxloss
Xmax = max(X,[],3) ;
ex = exp(bsxfun(@minus, X, Xmax)) ;

n = sz(1)*sz(2) ;
if nargin <= 2
  t = Xmax + log(sum(ex,3)) - reshape(X(c_), [sz(1:2) 1 sz(4)]) ;
  Y = sum(t(:)) / n ;
else
  Y = bsxfun(@rdivide, ex, sum(ex,3)) ;
  Y(c_) = Y(c_) - 1;
  Y = Y * (dzdy / n) ;
end

 As in the above code it is not divided by sz(4) which is the batch size. Am I missing something?

**Comment by vinhngx** [comment]
I notice that when using matconvnet for regression, it seems to struggle with even a simple data set like

dim=10;
n=1000;
X=5*rand(n,dim);
y=X(:,1)+randn(n,1)+X(:,2);

f=1/10;
net.layers = {} ;
layerSize=500;
dropout=0

net.layers{end+1} = struct('type', 'conv', ...
        'weights', {{f*randn(1,1,dim,layerSize, 'single'),  zeros(1,layerSize,'single')}}, ...
        'stride', 1, ...
        'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;%last layer has soft-max
if dropout net.layers{end+1} = struct('type', 'dropout','rate',0.5) ;end;

net.layers{end+1} = struct('type', 'conv', ...
        'weights', {{f*randn(1,1,layerSize,1, 'single'),  zeros(1,1,'single')}}, ...
        'stride', 1, ...
        'pad', 0) ;
net.layers{end+1} = struct('type', 'squaredloss') ;


opts.train.gpus = [] ;
opts.train.numEpochs=100;
opts.train.momentum=0.1;
opts.train.learningRate=1e-3;
opts.train.batchSize = 10000;

imdb.images.data=single(reshape(X',1,1,dim,n));
imdb.images.labels=y';
imdb.images.set=[ones(1,round(2*n/3)) 2*ones(1,n-round(2*n/3))];
imdb.images.set=imdb.images.set(randperm(n)); %randomly shuffle train-test data

[net, info] = cnn_train_regression(net, imdb, @getBatch, opts.train,  'val', find(imdb.images.set == 2)) ;


Playing with different initialization and learning rate, it still takes like hundreds of iterations to converge. 
Does anyone have the same experience? @lenck @ByungtaeAhn @AnanS
